pkg/handler/iam_handler_test.go

```diff
@@ -0,0 +1,563 @@
+// Copyright 2023 The Authors (see AUTHORS file)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// handler package that handles IAM request and IAM policy updates.
+package handler
+
+import (
+	"context"
+	"fmt"
+	"testing"
+	"time"
+
+	"cloud.google.com/go/iam/apiv1/iampb"
+	"cloud.google.com/go/resourcemanager/apiv3/resourcemanagerpb"
+	"github.com/abcxyz/access-on-demand/apis/v1alpha1"
+	"github.com/abcxyz/pkg/testutil"
+	"github.com/google/go-cmp/cmp"
+	"github.com/sethvargo/go-retry"
+	"google.golang.org/api/option"
+	"google.golang.org/genproto/googleapis/type/expr"
+	"google.golang.org/grpc"
+	"google.golang.org/protobuf/testing/protocmp"
+
+	resourcemanager "cloud.google.com/go/resourcemanager/apiv3"
+)
+
+func TestDo(t *testing.T) {
+	t.Parallel()
+
+	now := time.Now().UTC()
+	cases := []struct {
+		name                    string
+		organizationsServer     *fakeServer
+		foldersServer           *fakeServer
+		projectsServer          *fakeServer
+		request                 *v1alpha1.IAMRequestWrapper
+		wantPolicies            []*v1alpha1.IAMResponse
+		wantErrSubstr           string
+		wantOrganizationsPolicy *iampb.Policy
+		wantFoldersPolicy       *iampb.Policy
+		wantProjectsPolicy      *iampb.Policy
+	}{
+		{
+			name: "success",
```

Comment:
Do you mind splitting this case up? 1. happy path 2. ignore unrelated roles 3. ignore non-AOD bindings 4. clean up duplicated members. This would us to easily identify which case we failed if we changed any logic.